### PR TITLE
Add wick T1234 for xcf

### PR DESCRIPTION
**Pull request by londumas**
This PR adds the "wick" decomposition for the diagram T1,T2,T3 and T4 for the cross-correlation (see figure A.1 of https://arxiv.org/pdf/1708.02225.pdf).
It works differently than `do_wick` in the way that it reads the 1d auto-correlation directly from the matrix in the 2nd HDU of `do_cf1d` output ( xi1d (ll1,ll2) ). This should be faster and is more natural to get all the complexity of the matrix.
Its output has a `CO` keyword that allows to be used directly into `export.py`.
It also saves the different contribution of all the diagrams.
@ngbusca, I am not sure it takes too much more time to save all diagrams, what do you think? The main time consumption is in adding `if` statements I think.

Here is a plot comparing `N_pairs * diag(Cov)` for wick and sub-sampling for the LYA-QSO cross-correlation, for `--rej 0.98`:

capture d ecran de 2018-03-13 15-19-33

### CODE REVIEW

**Comment by londumas** [comment]
I also removed the `metal_grid` computation which is now useless.
If one day we need to get it back we can get it from the git history

**Comment by ngbusca** [comment]
@londumas this is great to have, thanks!! Would it be difficult to add T5/T6 as well?

**Comment by ngbusca** [comment]
>  I am not sure it takes too much more time to save all diagrams, what do you think? The main time consumption is in adding if statements I think.

I don't understand this question

**Comment by londumas** [comment]
@ngbusca, To add T5 and T6 we need to add the 3d auto-correlation of pixels, and we need to work on the code. Let's merge this PR and when needed, we can work on adding T5 and T6 in an efficient way.

You can forget the question.

**Comment by londumas** [comment]
@ngbusca, ok to merge?

**Comment by ngbusca** [comment]
@londumas almost, should we add a unittest?

**Comment by ngbusca** [comment]
@londumas can you post a comparisson of a fit done with Wick and with subsampling?

**Comment by ngbusca** [comment]
@londumas can you take a look at the Codacy review?

**Comment by londumas** [comment]
@ngbusca, your codacy is very nice. Have you tried running it on the master?
I am sure we have a lot of things to look at. For example codacy doesn't like `assert` because apparently they can get ignored

**Comment by londumas** [comment]
@ngbusca, do you want a unittest? I have no strong objection. I can build one and see if it is very long.

**Comment by ngbusca** [comment]
> @ngbusca, your codacy is very nice. Have you tried running it on the master?

Yes, there are hundreds of style changes to make!

**Comment by londumas** [comment]
> Yes, there are hundreds of style changes to make!

lol, so much fun!

**Comment by ngbusca** [comment]
> @ngbusca, do you want a unittest? I have no strong objection. I can build one and see if it is very long.

I don't know, Travis is already quite long... of course we don't want to inadvertidely break this functionality... I don't know what's best...

**Comment by londumas** [comment]
@ngbusca, Actually if you look, the Travis test is long mostly because of setting up the environment, not because of the unittest

**Comment by ngbusca** [comment]
> @ngbusca, Actually if you look, the Travis test is long mostly because of setting up the environment, not because of the unittest

That's true, I'll see if installing things with conda makes it faster

**Comment by londumas** [comment]
@ngbusca, I get `chi2 = 3215.48` for sub-sampling and `chi2 =3323.00` for Wick.
However, the best fit values and precision of the parameters don't change that much.
My guess is that it could come from:
- the missing T5 that has more contribution than T3+T4
- The sub-sampling presents spikes that resembles very much the absolute value of the cross-correlation. Maybe our calculation of the "Wick" decomposition is missing a term.
- The scale `NPTOT/NPUSED` is incorrect 

capture d ecran de 2018-03-15 17-50-46

**Comment by londumas** [comment]
We can see the disagreement between "Wick" and sub-sampling here for different nside values.
The spikes are very well seen for `nside = 32` or `128`.

capture d ecran de 2018-03-15 18-14-40

**Comment by ngbusca** [comment]
@londumas I am not at all worried about this "disagreement". The subsampling covariance is noisy and, as you've showed, the chi2 depends on nside. In particular, the smoothing model for subsampling is not very well motivated.

What surprises me is that your Wick doesn't seem to have the extra variance at small separations. I almost sure I've seen Wick reproducing that before. Could you plot the ratio variance(Wick)/variance(subsampling)? Here's an old plot I made a long time a go, for reference:

image

**Comment by londumas** [comment]
capture d ecran de 2018-03-16 11-51-48

**Comment by londumas** [comment]
Here is the diagonal of the covariance matrix times the number of pairs for a shuffle of the (ra,dec) positions of the forests. It proves two things:
- the missing power in the first plot comes from T5. Here because we have shuffled the forest this term is exactly zero.
- the spikes in the first plot do come from the cross-correlation

capture d ecran de 2018-03-22 07-51-46

File: py/picca/xcf.py [hunk 1]
```diff
@@ -328,3 +259,94 @@ def metal_dmat(pix,abs_igm="SiII(1526)"):
...
+    ba = bt + nt*bp
+
+    for k1 in range(rp.size):
+        p1 = ba[k1]
+        i1 = idxPix[k1]
+        q1 = idxQso[k1]
+        wAll[p1]  += we[k1]
+        nb[p1]    += 1
+        T1[p1,p1] += cf1d[i1,i1]*(we[k1]**2)
+
```
[the remark this answers is not in the document]
Comment (londumas) [reply py/picca/xcf.py]: `cf1d[i1,i1]` is basically the mean of `1./w1[i1]`. It must be less noisy.
Also using your equation you neglect the weight of the pixel in a certain way, no?

[the remark this answers is not in the document]
Comment (ngbusca) [reply py/picca/xcf.py]: > cf1d[i1,i1] is basically the mean of 1./w1[i1]. It must be less noisy

But here you need to keep the pixel-to-pixel fluctuations in the variance. With your algorithm you are giving the same variance to very noisy pixels and very low noise pixels. I think it's less correct. I agree that my formula might be missing a few factors, but it's trivial to work them out. Do you want me to give them to you exactly so you can make a quick comparison?

[the remark this answers is not in the document]
Comment (ngbusca) [reply py/picca/xcf.py]: In fact, what I'm proposing is what is done for the auto-Wick (look at the version in cf.py, the one in wick.py is incomplete and has some obvious bugs)

**Comment by ngbusca** [comment]
@londumas any news on this branch?

**Comment by londumas** [comment]
@ngbusca, I had no time to improve it.
Maybe this week or in two weeks.

**Comment by ngbusca** [comment]
`<delta_i^2> \propto 1/w_i` so the expression can be simplified

**Comment by ngbusca** [comment]
@londumas any news here?

**Comment by londumas** [comment]
@ngbusca, no news there. The current version does the job, but I didn't take the time to have a look at your remarks.

**Comment by londumas** [comment]
I send the wick for the auto and the cross with:
```
do_wick.py
--in-dir ../Delta_LYA/Delta/
--out wick_cf_forest1_LYA_line1_LYA.fits.gz
--nside 32
--cf1d cf1d.fits.gz
--rej 0.98

do_xwick.py
--in-dir ../Delta_LYA/Delta/
--drq $HOME/Data/Catalogs_for_MGII_studies/cat_QSO.fits
--out xwick_forest1_LYA_line1_LYA_obj1_QSO.fits.gz
--nside 32
--cf1d cf1d.fits.gz
--rej 0.98
```
It doesn't look to bad for the cross, but really not perfect for the auto.
Maybe I shouldn't use a rejection factor so low.
Or maybe I introduced a bug in the auto, I really don't think so.
The following plot shows the variance of the auto and the cross, it is produced using the
following code. I compare the results to the sub-sampling covariance.

wick_auto_var

wick_cross_var

```

import fitsio
import scipy as sp
import matplotlib.pyplot as plt

def test_wick():
    cov = {}
    nb = {}

    nb_bin_cross = 2500
    h1 = fitsio.FITS('../Correlations_LYA/forFits_cf_forest1_LYA_line1_LYA.fits.gz')
    h2 = fitsio.FITS('wick_forest1_LYA_line1_LYA.fits.gz')
    nbT = 2
    #nb_bin_cross = 5000
    #h1 = fitsio.FITS('../Correlations_LYA/forFits_xcf_forest1_LYA_line1_LYA_obj1_QSO.fits.gz')
    #h2 = fitsio.FITS('../Correlations_LYA/xwick_forest1_LYA_line1_LYA_obj1_QSO.fits.gz')
    #nbT = 4

    cov['Sub'] = h1[1]['CO'][:]
    cov['Wick'] = h2[1]['CO'][:]
    for i in range(1,nbT+1):
        cov['T'+str(i)] = h2[1]['T'+str(i)][:]

    nb['Sub'] = h1[1]['NB'][:]
    nb['Wick'] = h2[1]['NB'][:]*h2[1].read_header()['NPALL']/h2[1].read_header()['NPUSED']
    for i in range(1,nbT+1):
        nb['T'+str(i)] = h2[1]['NB'][:]*h2[1].read_header()['NPALL']/h2[1].read_header()['NPUSED']

    for name,c in cov.items():
        plt.plot(sp.diag(c)*nb[name],label=name)
    plt.xlabel('bin index',fontsize=20)
    plt.ylabel('Caa*nb',fontsize=20)
    plt.title('Auto',fontsize=20)
    plt.legend()
    plt.grid()
    plt.show()
```

**Comment by ngbusca** [comment]
@londumas I'm attaching all plots that show that Wick works extremely well. They were not produced with picca but with a code that predated it, I'm attaching them to show the level of agreement that we should be getting. In particular, for the cross you really expect that Wick reproduce the structure at small separations.

**Auto**:

image

**Cross**

image

**Comment by ngbusca** [comment]
note that the binning for the cross used to be 5 Mpc/h instead of the current 4Mpc/h, which is why there are less bins.

**Comment by londumas** [comment]
@ngbusca, thanks these plots are going to be useful.
Did you try the do_wick.py of picca? Or you never checked that it was correct

**Comment by ngbusca** [comment]
@londumas I don't remember doing a systematic check like the one you are doing right now...

File: py/picca/cf.py [hunk 2]
```diff
@@ -558,12 +559,16 @@ def fill_t123(r1,r2,ang,w1,w2,z1,z2,c1d_1,c1d_2,w123,t123_loc,same_half_plate):
...
+        T1[ba[k],ba[k]] += we[k]/zw[k]
+
         for l in range(k+1,w.sum()):
             i2 = bins[l]%n1
             j2 = (bins[l]-i2)//n1
             prod = c1d_1[i1,i2]*c1d_2[j1,j2]
-            t123_loc[ba[k],ba[l]]+=prod
-            t123_loc[ba[l],ba[k]]+=prod
+            T2[ba[k],ba[l]] += prod
+            T2[ba[l],ba[k]] += prod
```
Comment (ngbusca) [inline py/picca/cf.py]: @londumas there's a problem here with the definition of `T1, T2, T3`. `T1` should contain only the product of the variances, `T2` a variance times a c1d and `T3` the product of two c1d. From what I see you have `T1` and `T3` but `T2` is missing.

File: py/picca/xcf.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ngbusca) [reply py/picca/xcf.py]: @londumas the `T1` term shouldn't contain any c1d but be a pure variance term. You can construct this from the weights `( = ((1+z)^something/(1+z0)^something)/var)` by dividing by the extra factors of `(1+z)^...`

File: py/picca/xcf.py [hunk 3]
```diff
@@ -328,3 +259,94 @@ def metal_dmat(pix,abs_igm="SiII(1526)"):
...
+            p2 = ba[k2]
+            i2 = idxPix[k2]
+            q2 = idxQso[k2]
+            wcorr = cf1d[i1,i2]*we[k1]*we[k2]
+            if q1==q2:
+                T2[p1,p2] += wcorr
+                T2[p2,p1] += wcorr
+            elif i1==i2:
+                T3[p1,p2] += wcorr
+                T3[p2,p1] += wcorr
```
[the remark this answers is not in the document]
Comment (ngbusca) [reply py/picca/xcf.py]: similar here, the `T2` should contain a factor of c1d but the `T3` should be proportional to the variance ( \propto 1/weight)

**Comment by londumas** [comment]
The current version of the branch fixed the different remarks of @ngbusca on the cross and auto-correlation. It adds computing all the diagrams involving one or two line-of-sights for auto (e.g. Lya x Lya), cross (e.g. Lya x qso) and cross-correlation between two different forests (ex: Lya(Lya) x Lya(Lyb)).
The higher order diagrams involving 3 or 4 line-of-sights will be implemented in a future PR. Let's work on merging this one.

@ngbusca, could you do a review?

The following plots show the contribution of the different diagrams and a comparison with sub-sampling for the cross and the auto.

cf_diag

cf_nodiag

cf_nodiag2

xcf_diag

xcf_nodiag

xcf_nodiag2

File: py/picca/cf.py (hunk 2 above)
Comment (londumas) [reply py/picca/cf.py]: @ngbusca, fixed.

File: py/picca/xcf.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (londumas) [reply py/picca/xcf.py]: @ngbusca, fixed.

File: py/picca/xcf.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (londumas) [reply py/picca/xcf.py]: @ngbusca, fixed.

**Comment by ngbusca** [comment]
@londumas thanks for the work, I'll take a look. In some of the plots above I don't see the "Wickall" orange line...

**Comment by londumas** [comment]
@ngbusca, Thanks. Sorry for the confusion, in the plots giving the diagonal, Ti gives the contribution of only the given diagram. In the plots giving the correlation matrix, Ti gives the contribution up to the given diagram. So we don't see `Wickall` in the correlation-matrix plots because they are behind T4=T1234=Wickall.

**Comment by londumas** [comment]
The results are given in the wiki page:
https://trac.sdss.org/wiki/eBOSS/lya/DR16_analysis/wick_covariance

**Comment by londumas** [comment]
@vserret, I understand you ran the wick covariance? It is good, because I need people to test that.
Could you make a review? And then I would merge.

**Comment by vserret** [comment]
There are still some comments " ### TODO work on the good formula" that should be removed.
I couldn't run the code since nersc has been very busy these last weeks

**Comment by londumas** [comment]
merging this PR, with partial reviews.

**Pull request merged**